process/smartContract/processorV2/vmOutputAccountsProcessor.go

```diff
@@ -0,0 +1,313 @@
+	sameCodeMetadata := bytes.Equal(outputAccountCodeMetadataBytes, stateAccount.GetCodeMetadata())
+	if sameCode && sameCodeMetadata {
+		return nil
+	}
+
+	currentOwner := stateAccount.GetOwnerAddress()
+	isCodeDeployerSet := len(outputAccount.CodeDeployerAddress) > 0
+	isCodeDeployerOwner := bytes.Equal(currentOwner, outputAccount.CodeDeployerAddress) && isCodeDeployerSet
+
+	noExistingCode := len(oap.sc.accounts.GetCode(stateAccount.GetCodeHash())) == 0
+	noExistingOwner := len(currentOwner) == 0
+	currentCodeMetadata := vmcommon.CodeMetadataFromBytes(stateAccount.GetCodeMetadata())
+	newCodeMetadata := vmcommon.CodeMetadataFromBytes(outputAccountCodeMetadataBytes)
+	isUpgradeable := currentCodeMetadata.Upgradeable
+	isDeployment := noExistingCode && noExistingOwner
+	isUpgrade := !isDeployment && isCodeDeployerOwner && isUpgradeable
+
+	entry := &vmcommon.LogEntry{
+		Address: stateAccount.AddressBytes(),
+		Topics: [][]byte{
+			outputAccount.Address, outputAccount.CodeDeployerAddress,
+		},
+	}
+
+	if isDeployment {
+		// At this point, we are under the condition "noExistingOwner"
+		stateAccount.SetOwnerAddress(outputAccount.CodeDeployerAddress)
+		stateAccount.SetCodeMetadata(outputAccountCodeMetadataBytes)
+		stateAccount.SetCode(outputAccount.Code)
+		log.Debug("updateSmartContractCode(): created", "address", oap.sc.pubkeyConv.Encode(outputAccount.Address), "upgradeable", newCodeMetadata.Upgradeable)
+
+		entry.Identifier = []byte(core.SCDeployIdentifier)
+		vmOutput.Logs = append(vmOutput.Logs, entry)
+		return nil
+	}
+
+	if isUpgrade {
+		stateAccount.SetCodeMetadata(outputAccountCodeMetadataBytes)
+		stateAccount.SetCode(outputAccount.Code)
+		log.Debug("updateSmartContractCode(): upgraded", "address", oap.sc.pubkeyConv.Encode(outputAccount.Address), "upgradeable", newCodeMetadata.Upgradeable)
+
+		entry.Identifier = []byte(core.SCUpgradeIdentifier)
+		vmOutput.Logs = append(vmOutput.Logs, entry)
+		return nil
+	}
+
+	return nil
+}
+
+func (oap *VMOutputAccountsProcessor) updateAccountNonceIfThereIsAChange(
+	acc state.UserAccountHandler,
+	outAcc *vmcommon.OutputAccount) error {
+	if !oap.processNonce {
+		return nil
+	}
+	if outAcc.Nonce != acc.GetNonce() && outAcc.Nonce != 0 {
+		if outAcc.Nonce < acc.GetNonce() {
+			return process.ErrWrongNonceInVMOutput
+		}
+
+		nonceDifference := outAcc.Nonce - acc.GetNonce()
+		acc.IncreaseNonce(nonceDifference)
+	}
+	return nil
+}
+
+func (oap *VMOutputAccountsProcessor) updateAccountBalanceStep(
+	acc state.UserAccountHandler,
+	outAcc *vmcommon.OutputAccount,
+	sumOfAllDiff *big.Int) (*big.Int, error) {
+	if !oap.processBalanceDelta {
+		return sumOfAllDiff, nil
+	}
+	// if no change then continue to next account
+	if outAcc.BalanceDelta == nil || outAcc.BalanceDelta.Cmp(zero) == 0 {
+		err := oap.sc.accounts.SaveAccount(acc)
+		if err != nil {
+			return sumOfAllDiff, err
+		}
+
+		return sumOfAllDiff, nil
+	}
+
+	sumOfAllDiff = sumOfAllDiff.Add(sumOfAllDiff, outAcc.BalanceDelta)
```

Comment:
this will actually change the provided pointer, if the code slightly change, it might hide hideous bugs. Suggestion to change to:
```GO
sumOfAllDiff = big.NewInt(0).Add(sumOfAllDiff, outAcc.BalanceDelta)
```

Reply:
fixed